### PR TITLE
Implement CSI SnapshotMetadata service

### PR DESCRIPTION
**Pull request by PrasadG193**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR implements SnapshotMetadata service from the [CSI specs](https://github.com/container-storage-interface/spec/blob/master/csi.proto#L130). This service in combination with [external-snapshot-metadata](https://github.com/kubernetes-csi/external-snapshot-metadata) will be used to compare and query changed block metadata between two snapshots.
Right now, the service will be disabled by default till `external-snapshot-metadata` alpha is release.

The changes in deploy script to add `external-snapshot-metadata` sidecar will be added after alpha release.

#### NOTE
This is a sample implementation of SnapshotMetadata intended to be used for demo and CI testing purpose only.
This should not be used in production or as an example about how to write a real driver.

##### Test plan
Complete test plan is documented at - https://gist.github.com/PrasadG193/1fab082f03bbdb7db25a41ed17b7e7c1

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add sample implementation of CSI SnapshotMetadata service 
```

### CODE REVIEW

File: pkg/hostpath/snapshotmetadata.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	source, target, err := openFiles(sourcePath, targetPath)
+	if err != nil {
+		return err
+	}
+	defer source.Close()
+	defer target.Close()
+
+	if err := seekToOffset(source, target, startingOffset); err != nil {
+		return err
+	}
```
Comment (carlbraganza) [inline pkg/hostpath/snapshotmetadata.go:59]: This is a gRPC `OUT_OF_RANGE` error that must be forwarded to the client.

File: pkg/hostpath/snapshotmetadataserver.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+		return status.Error(codes.InvalidArgument, "Source volume does not have block mode access type")
+	}
+
+	allocatedBlocks := make(chan []*csi.BlockMetadata, 100)
+	go func() {
+		err := hp.getAllocatedBlockMetadata(ctx, hp.getSnapshotPath(snapID), req.StartingOffset, state.BlockSizeBytes, req.MaxResults, allocatedBlocks)
+		if err != nil {
+			klog.Errorf("failed to get allocated block metadata: %v", err)
+		}
+	}()
```
Comment (carlbraganza) [inline pkg/hostpath/snapshotmetadataserver.go:60]: The error returned by the function may be significant and require a specific gRPC code - e..g. `OUT_OF_RANGE`.

I suggest not using a goroutine but instead us a synchronous I/O access pattern - i.e. equivalent to`Open`, `Seek`, `Read` and `Close`.

File: pkg/hostpath/snapshotmetadataserver.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	// Check arguments
+	snapID := req.GetSnapshotId()
+	if len(snapID) == 0 {
+		return status.Error(codes.InvalidArgument, "SnapshotID missing in request")
+	}
+
+	// Load snapshots
+	source, err := hp.state.GetSnapshotByID(snapID)
+	if err != nil {
+		return status.Error(codes.Internal, "Cannot find the snapshot")
```
Comment (carlbraganza) [inline pkg/hostpath/snapshotmetadataserver.go:40]: This should be `codes.NotFound`.

File: pkg/hostpath/snapshotmetadata.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+import (
+	"bytes"
+	"io"
+	"os"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+	"k8s.io/klog/v2"
+)
+
```
Comment (xing-yang) [inline pkg/hostpath/snapshotmetadata.go:28]: Can you add a note here that this is a sample implementation, and not for production use.

**Comment by xing-yang** [comment]
This is a new feature. Please add a release note.
Can you add a note in the PR description that this is a sample implementation, and not for production use.

**Comment by PrasadG193** [comment]
@carlbraganza @xing-yang I've addressed the feedback. PTAL

**Comment by carlbraganza** [comment]
@PrasadG193 LGTM!  Need approval from @xing-yang or @bswartz

File: pkg/hostpath/snapshotmetadataserver.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+		if cbErr != nil {
+			if errors.Is(cbErr, context.Canceled) {
+				klog.V(4).Info("context canceled while getting allocated block metadata, returning")
+				return nil
+			}
+			if errors.Is(cbErr, context.DeadlineExceeded) {
+				klog.V(4).Info("context deadline exceeded while getting allocated block metadata, returning")
+				return nil
+			}
+			if err == io.EOF {
```
Comment (iPraveenParihar) [inline pkg/hostpath/snapshotmetadataserver.go:86]:
```suggestion
			if errors.Is(cbErr, io.EOF) {
```

File: pkg/hostpath/snapshotmetadataserver.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+		if cbErr != nil {
+			if errors.Is(cbErr, context.Canceled) {
+				klog.V(4).Info("context canceled while getting changed block metadata, returning")
+				return nil
+			}
+			if errors.Is(cbErr, context.DeadlineExceeded) {
+				klog.V(4).Info("context deadline exceeded while getting changed block metadata, returning")
+				return nil
+			}
+			if err == io.EOF {
```
Comment (iPraveenParihar) [inline pkg/hostpath/snapshotmetadataserver.go:173]:
```suggestion
			if errors.Is(cbErr, io.EOF) {
```

**Review by iPraveenParihar — CHANGES_REQUESTED** [review]

**Comment by xing-yang** [comment]
Can you add your testing steps?

**Comment by Rakshith-R** [comment]
> Can you add your testing steps?

hey @xing-yang ,

@PrasadG193 is on vacation and will be back thursday.

He had developed a client to test the new RPCs https://github.com/PrasadG193/external-snapshot-metadata-client 
which has been adopted as a tool in this pr https://github.com/kubernetes-csi/external-snapshot-metadata/pull/64 

We are creating a gh action that adds simple e2e
https://github.com/kubernetes-csi/external-snapshot-metadata/issues/67 
```
Add a github action that:
- checkout repository
- sets up minikube
- builds latest sidecar image
- deploys csi-driver-hostpath driver resources
- runs simple e2e tests to create pvc, pod and snapshots and to write data.
- run client to verify working of both RPCs.
```

Sample run https://github.com/Rakshith-R/external-snapshot-metadata/actions/runs/12163453092/job/33922692908
(with the fix @iPraveenParihar has suggested above)

File: pkg/hostpath/snapshotmetadataserver.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+		return status.Error(codes.Unavailable, fmt.Sprintf("snapshot %v is not yet ready to use", snapID))
+	}
+
+	vol, err := hp.state.GetVolumeByID(source.VolID)
+	if err != nil {
+		return err
+	}
+	if vol.VolAccessType != state.BlockAccess {
+		return status.Error(codes.InvalidArgument, "source volume does not have block mode access type")
+	}
```
Comment (carlbraganza) [inline pkg/hostpath/snapshotmetadataserver.go:55]: @PrasadG193 Is this really necessary? Most CSI drivers that make volume snapshots do so regardless of the access mode. It imposes an additional hurdle for testing.

Comment (Rakshith-R) [reply pkg/hostpath/snapshotmetadataserver.go:55]: volume mode can be converted from filesystem to block ex:
filesystem parent PVC -> snapshot -> annotate VSC to permit conversion of volumeMode -> restore into block mode PVC.

https://kubernetes-csi.github.io/docs/prevent-volume-mode-conversion.html

Backup solution will use this route to backup filesystem PVCs mounted on block devices provided by storage solutions.

**Comment by PrasadG193** [comment]
@xing-yang I have documented complete test plan along with steps to deploy csi driver with external-snapshot-metadata here - https://gist.github.com/PrasadG193/1fab082f03bbdb7db25a41ed17b7e7c1

File: pkg/hostpath/snapshotmetadataserver.go (hunk 7 above)
Comment (PrasadG193) [reply pkg/hostpath/snapshotmetadataserver.go:55]: thanks, I will test and address that

Comment (Rakshith-R) [reply pkg/hostpath/snapshotmetadataserver.go:55]: > volume mode can be converted from filesystem to block ex: filesystem parent PVC -> snapshot -> annotate VSC to permit conversion of volumeMode -> restore into block mode PVC.
> 
> https://kubernetes-csi.github.io/docs/prevent-volume-mode-conversion.html
> 
> Backup solution will use this route to backup filesystem PVCs mounted on block devices provided by storage solutions.

We discussed this offline.

Hostpath driver creates a directory for a filesystem PVC
and creates a block device for block PVC.
https://github.com/kubernetes-csi/csi-driver-host-path/blob/2211f6ddc9ebf9cb23a3f29021ccc55b63e8f523/pkg/hostpath/hostpath.go#L185-L204

Hence, we cannot convert a directory part of a filesystem into block device.

**Comment by iPraveenParihar** [comment]
LGTM! @PrasadG193

**Review by bswartz — CHANGES_REQUESTED** [review]
Overall this looks pretty good. There are some inefficient bits of code but for a proof-of-concept it seems fine. I found one bug with error handling.

File: pkg/hostpath/snapshotmetadata.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+			return err
+		}
+	}
+	return nil
+}
+
+func openFiles(basePath, targetPath string) (base, target *os.File, err error) {
+	target, err = os.Open(targetPath)
+	if err != nil {
+		base.Close()
```
[the remark this answers is not in the document]
Comment (PrasadG193) [reply pkg/hostpath/snapshotmetadata.go:94]: AAh, I missed moving the close to L102 after refactoring the code. I've fixed this now

**Review by bswartz — APPROVED** [review]
/lgtm

File: cmd/hostpathplugin/main.go [hunk 9]
```diff
@@ -106,6 +110,14 @@ func main() {
 		cfg.MaxVolumeExpansionSizeNode = cfg.MaxVolumeSize
 	}
 
+	// validate snapshot-metadata-type arg block type
+	bt, ok := csi.BlockMetadataType_value[*snapshotMetadataBlockType]
+	if !ok {
+		fmt.Printf("invalid snapshot-metadata-block-type passed, please pass one of the - FIXED_LENGTH, VARIABLE_LENGTH")
```
Comment (pnhoang75) [inline cmd/hostpathplugin/main.go:116]: nit: use upper case for beginning of the error/exit message to conform with other error messages in the file.

**Review by bswartz — APPROVED** [review]
/lgtm

**Pull request merged**